File: server/c/glob/Threadable.cpp
/* ***************************************************************************

   Copyright (c): 2013 Hericus Software, Inc.

   License: The MIT License (MIT)

   Authors: Steven M. Cherry

*************************************************************************** */

#include "TheMain.h"
#include "Threadable.h"
using namespace Helix::Glob;

#include "AnException.h"
#include "EnEx.h"
using namespace SLib;

#include <stdexcept>

Threadable::Threadable()
{
	EnEx ee(FL, "Threadable::Threadable()");

	m_state = Threadable::Init;
	m_shutdown = false;
	m_reread_config = false;

	// This should be re-set by each child class
	m_descr = "Generic Threadable";

}

Threadable::Threadable(const Threadable& c)
{
	EnEx ee(FL, "Threadable::Threadable(const Threadable& c)");

	m_state = Threadable::Init;
	m_shutdown = false;
	m_reread_config = false;
	m_descr = c.m_descr;

}

Threadable& Threadable::operator=(const Threadable& c)
{
	EnEx ee(FL, "Threadable::operator=(const Threadable& c)");

	m_state = c.m_state;
	m_shutdown = c.m_shutdown;
	m_reread_config = c.m_reread_config;
	m_descr = c.m_descr;

	return *this;
}

Threadable::~Threadable()
{
	EnEx ee(FL, "Threadable::~Threadable()");
	m_state = Threadable::Done;
	m_shutdown = true;
}

void Threadable::PreExecute(void)
{
	EnEx ee(FL, "Threadable::PreExecute()");

	// Set our state
	m_state = Threadable::Startup;

	// Add ourselves to the global thread list
	TheMain::getInstance()->GetThreadList().AddThreadable(this);	

}

void Threadable::PostExecute(void)
{
	EnEx ee(FL, "Threadable::PostExecute()");

	// Set our state
	m_state = Threadable::Finishing;

	// Remove ourselves from the global thread list
	TheMain::getInstance()->GetThreadList().RemoveThreadable(this);

}

int Threadable::Status(void)
{
	EnEx ee(FL, "Threadable::Status()");
	return m_state;
}

void Threadable::Shutdown(void)
{
	EnEx ee(FL, "Threadable::Shutdown()");

	m_shutdown = true;

}

void Threadable::ReReadConfig(void)
{
	EnEx ee(FL, "Threadable::ReReadConfig()");

	m_reread_config = true;

}

twine& Threadable::Description(void)
{
	EnEx ee(FL, "Threadable::Description()");
	return m_descr;
}

void Threadable::ThreadID(THREAD_ID_TYPE tid)
{
	EnEx ee(FL, "Threadable::ThreadID(THREAD_ID_TYPE tid)");
	m_tid = tid;
}

THREAD_ID_TYPE Threadable::ThreadID(void)
{
	EnEx ee(FL, "Threadable::ThreadID()");
	return m_tid;
}

void* Threadable::Start(void* arg)
{
	EnEx ee(FL, "Threadable::Start()", true);

	// Cast the argument
	Threadable* t = (Threadable*)arg;

	// Set our thread id in the threadable object:
	t->ThreadID(CURRENT_THREAD_ID);
	
	// If PreExecute fails, then don't do anything else.
	// If PreExecute goes well, then guarantee to call
	// PostExecute, even if an exception happens in Execute().
	
	try {
		t->PreExecute();
	} catch (AnException& e){
		ERRORL(FL, "Exception during PreExecute: %s", e.Msg());
		return (void*)NULL;
	} catch (std::exception& stde){
		ERRORL(FL, "STD::Exception during PreExecute: %s", stde.what() );
		return (void*)NULL;
	} catch (...){
		ERRORL(FL, "Unknown Exception during PreExecute.");
		return (void*)NULL;
	}

	try {
		t->Execute();
	} catch (AnException& e) {
		ERRORL(FL, "Exception during Execute: %s", e.Msg());
	} catch (std::exception& stde){
		ERRORL(FL, "STD::Exception during Execute: %s", stde.what() );
	} catch (...){
		ERRORL(FL, "Unknown Exception during Execute");
	}

	try {
		t->PostExecute();
	} catch (AnException& e){
		ERRORL(FL, "Exception during PostExecute: %s", e.Msg());
	} catch (std::exception& stde){
		ERRORL(FL, "STD::Exception during PostExecute: %s", stde.what() );
	} catch (...){
		ERRORL(FL, "Unknown Exception during PostExecute");
	}

	return (void*)NULL;
}

File: server/net/glob/Threadable.cs
/* ***************************************************************************

   Copyright (c): 2014 Hericus Software, Inc.

   License: The MIT License (MIT)

   Authors: Steven M. Cherry

*************************************************************************** */

using System;

namespace Helix.Glob
{
	public abstract class Threadable
	{
		public const int Init = 0;
		public const int Startup = 1;
		public const int Running = 2;
		public const int Sleeping = 3;
		public const int Finishing = 4;
		public const int Done = 5;

		public Threadable ()
		{
			m_state = Threadable.Init;
			m_shutdown = false;
			m_reread_config = false;

			// This should be re-set by each child class
			m_descr = "Generic Threadable";
		}

		/**
		  * This is how the caller let's us know what our thread ID
		  * is.  This is typically only called by Threadable::Start().
		  * It shouldn't be necessary to be called by anyone else.
		  */
		public void ThreadID(int tid)
		{
			m_tid = tid;
		}

		/**
		  * This is how others ask us about our thread ID.  We return
		  * whatever was set by the initial call to ThreadID(...);
		  */
		public int ThreadID()
		{
			return m_tid;
		}

		/**
		  * This method is called before execution of
		  * the main body of the thread.  Subclasses may
		  * override this if necessary.  If they do, they should
		  * still ensure that this gets executed by doing:
		  * Threadable::PreExecute() at the top of their version
		  * of this method.
		  */
		public virtual void PreExecute()
		{
			using(EnEx ee = new EnEx("Threadable::PreExecute()")){
				// Set our state
				m_state = Threadable::Startup;

				// Add ourselves to the global thread list
				TheMain.getInstance().GetThreadList().AddThreadable(this);	
			}
		}

		/**
		  * This is the main body of the thread.  This must
		  * be implemented by the subclass.
		  */
		public abstract void Execute();

		/**
		  * This method is called after execution of
		  * the main body of the thread.  Subclasses may
		  * override this if necessary.  If they do, they should
		  * still ensure that this gets executed by doing:
		  * Threadable::PostExecute() at the top of their version
		  * of this method.
		  */
		public virtual void PostExecute()
		{
			using(EnEx ee = new EnEx("Threadable::PostExecute()")){
				// Set our state
				m_state = Threadable.Finishing;

				// Remove ourselves from the global thread list
				TheMain.getInstance().GetThreadList().RemoveThreadable(this);
			}
		}

		/**
		  * This method will return the current state
		  * of this thread.  It will be one of the
		  * above const int values. (Init, Startup,
		  * Running, Sleeping, Finishing, Done).
		  */
		public int Status()
		{
			return m_state;
		}

		/**
		  * This method will tell the thread to stop
		  * processing and shutdown gracefully.
		  */
		public void Shutdown()
		{
			using(EnEx ee = new EnEx("Threadable::Shutdown()")){
				m_shutdown = true;
			}
		}

		/**
		  * This method will tell the thread to re-read it's config
		  * info from TheMain.
		  */
		public void ReReadConfig()
		{
			using(EnEx ee = new EnEx("Threadable::ReReadConfig()")){

			}
		}

		/**
		  * This method returns a description of what
		  * the point of this threadable object is.
		  * i.e. what it is trying to do.
		  */
		public string Description()
		{
			return m_descr;
		}

		/**
		  * This static method should be used as the
		  * argument to the Thread::start() routine
		  * to start a Threadable object.  The arg
		  * value is expected to be a pointer to a 
		  * Threadable object.
		  * <P>
		  * Using this method ensures that PreExecute, Execute,
		  * and PostExecute all get called correctly, even in the
		  * case of exceptions.
		  */
		public static void Start(object arg)
		{
			using(EnEx ee = new EnEx("Threadable::Start()")){
				// Cast the argument
				Threadable t = (Threadable)arg;

				// Set our thread id in the threadable object:
				t.ThreadID(System.Threading.Thread.CurrentThread.ManagedThreadId);

				// If PreExecute fails, then don't do anything else.
				// If PreExecute goes well, then guarantee to call
				// PostExecute, even if an exception happens in Execute().

				try {
					t.PreExecute();
				} catch (Exception e){
					Log.Error("Exception during PreExecute: {0}", e.Message);
					return;
				}

				try {
					t.Execute();
				} catch (Exception e) {
					Log.Error("Exception during Execute: {0}", e.Message);
				}

				try {
					t.PostExecute();
				} catch (Exception e){
					Log.Error("Exception during PostExecute: {0}", e.Message);
				}

				return;
			}
		}

		/// Our current thread state variable.
		protected volatile int m_state;

		/// Our shutdown flag
		protected volatile bool m_shutdown;

		/// Our re-read config flag
		protected volatile bool m_reread_config;

		/// A description of what this object does
		protected string m_descr;

		/// Our thread id
		protected int m_tid;

	}
}

